### ISSUE TITLE
Can I or will I be able to run VSCode on Surface Pro X?

### ISSUE THREAD

**Issue by vilic** [microsoft/vscode#81854]
I use Remote SSH most of the time btw.

**Comment by FilipeRamalho** [comment]
This seems to be related to #33620
And there is already an x86 version of VSCode which should run on the Surface Pro X through x86 emulation.

**Comment by vilic** [comment]
@FilipeRamalho thanks for the reference. I am a little bit worry about the performance stuff. Can I rely on Surface Pro X to be my daily PC for development?

**Comment by FilipeRamalho** [comment]
@vilic Without going too much off topic. I am also considering the Pro X and thinking about the use of development tools on it. As it releases on November there is enough time to see if it's possible to use as an daily PC for development. 
I can run VS Code on my SP4-m3, so it should work fine on the Pro X, but other tools like VS or IntelliJ may not work fine.

**Comment by hernancmartinez** [comment]
@FilipeRamalho  I just pre-ordered mine.

**Comment by magic-k** [comment]
@vilic, @FilipeRamalho  If you build vscode by yourself or get a arm64 version from a third party site you get code-oss. One of the major differences between official builds and open source builds is that the remote extensions refuse to work with oss. So in your case you will have to work in the x86 emulation. Which is probably currently the best way as the c# and c++ extensions also don't work on arm64. Probably a lot other extensions, too. Although i'm from the linux side i'm hoping that with the surface pro x native arm gets a bit of a boost.

**Comment by FilipeRamalho** [comment]
@magic-k I think that code-oss is a proof of concept, so we know that vs code itself can work on win-arm64.
It would be super cool if the linux community supports the Pro X to boost native ARM.

**Comment by magic-k** [comment]
@FilipeRamalho  I was more thinking the other way around. Microsoft forces it's own vscode department to support a native arm64 version, forcing the extensions to follow. From there a native linux version is probably only a short step. code-oss is the open source community portion of vscode, which works on all platforms you can compile it for including linux arm64 but without proprietary microsoft parts and native x86 parts.
Yes, these extensions contain native x86 components.

**Comment by FilipeRamalho** [comment]
@magic-k The VSCode community will hopefully follow.

**Comment by brion** [comment]
The primary trouble I have using the x86 VS Code build on an ARM64 Win10 laptop is that the remote extensions don't support ARM64 Linux (it seems to download x86-64 code which then fails to run?), so WSL-based development isn't as smooth as it could be.

Performance of the editor is adequate for my purposes.

**Comment by FilipeRamalho** [comment]
@brion The extensions seem to be the primary problem when using VSCode are extensions with X64 code, I'm no that well-versed into Code extensions, but I will try to find out what is been compiled into x64 and if it can be compiled into x86 or even arm64.

**Comment by masknu** [comment]
I'm also thinking about using Pro X to develop Golang backend applications.

**Comment by magic-k** [comment]
As long as everything runs in the emulation, it'll probably works. delve, the golang debugger won't work on arm64.

**Comment by voronoipotato** [comment]
Firstly the original blocking issue has been resolved, there is official windows ARM64 electron support, hooray :tada: .  

I'm using a surface pro X now, the x86 version of code does work, but there's slightly disruptive visual lag on most actions, though typing seems to be fine. I'm also a mostly remote use person so I'm not dying here but it seems like it should be low hanging fruit.

**Comment by headmelted** [comment]
I'm happy to try adding this to https://github.com/headmelted/codebuilds now that Electron supports Windows ARM64.

Is there interest in this, given that it's not an official build so you would need to accept that there may be inconsistencies in terms of what's support from an extension perspective.

**Comment by voronoipotato** [comment]
Thank you @headmelted !!! I can and will gracefully accept whatever extension challenges may come.

**Comment by JimPriestley** [comment]
> The primary trouble I have using the x86 VS Code build on an ARM64 Win10 laptop is that the remote extensions don't support ARM64 Linux (it seems to download x86-64 code which then fails to run?), so WSL-based development isn't as smooth as it could be.
> 
> Performance of the editor is adequate for my purposes.

This was fixed in the November 2019 1.41 release. The x86 Windows UI will now correctly use the x64 ARM vscode-server in WSL.

**Comment by JimPriestley** [comment]
The Surface ProX is a great machine for on the go connectivity, I have a larger machine for work at my desk. I would love to see the UI deployed as ARM64 on Windows. I feel that skipping the x86 emulation would increase performance and potentially decrease Electron's power consumption.

**Comment by chrisdias** [comment]
What types of apps do you want to build on the machine? What are the target runtimes (e.g. Node, Go, Rust, Python, etc.). Are there ARM64 versions of those tools?  Or, are you primarily doing remote development (e.g. SSH) to another machine?

**Comment by JimPriestley** [comment]
I primarily work in C# Core, Python and Azure CLI.

My desired setup on the Surface ProX is to work both local WindowsARM and in WSL2 Remoting in native ARM64.

I understand from the thread that the C# plugins have x86 dependencies, but I expect that no plug in maintainers will consider updating to native ARM on Win/Linux support until the IDE supports the scenario.

**Comment by headmelted** [comment]
To add to Chris’ point above and pre-empt the next question, some of the blanks in tooling probably can be filled in with emulation - but from experience I’d expect very mixed results, and in any case it could get so messy so quickly that it becomes more hassle than it’s worth for an end user (example: x86 on ARM targeting x64 would be three triplets to track in GCC). Even if the tools you need work in an emulator (which is not guaranteed), performance would likely be very poor.

**Comment by voronoipotato** [comment]
It's incumbent on the host to lead the push. If the host doesn't support it there's absolutely no reason for the client to support it. The idea that clients will work to support something that the host doesn't is fantasy.

**Comment by JimPriestley** [comment]
> What types of apps do you want to build on the machine? What are the target runtimes (e.g. Node, Go, Rust, Python, etc.). Are there ARM64 versions of those tools?  Or, are you primarily doing remote development (e.g. SSH) to another machine?

arm64 Linux versions of DotNet 3.1, AZCLI, Docker-ce and Python do exist, and work well in the WSL Remoting on the Surface ProX. As stated above, running VSCode Shell on ARM64 on Windows would be a first step in supporting Windows on ARM as a dev work environment, and is a pre-requisite for any plugin authors then also supporting Win on ARM natively.

Edge on Chromium for WinARM is now in Dev, and is supposed to release on January 15th, according to news stories.

Other threads indicate there is now an Electron release that supports Win on ARM64.

**Comment by ryan-allen** [comment]
The minute there is an official build for ARM, I'm buying a Surface Pro X.

The SSH and WSL remoting is awesome, I can't wait to be able to use it as a development machine!

**Comment by JimPriestley** [comment]
I have a Surface ProX. 32-bit Windows install of Code now runs perfectly well, including WSL-Remote.

I prefer WSL2 over WSL1, so I am using the Windows Insider Slow Ring on the ProX, with current GA VsCode, and ARM64 Ubuntu in WSL2.

**Comment by ryan-allen** [comment]
@JimPriestley Oh... neat! Thanks for the insight... I think I'm brave enough to make the jump now! :)

**Comment by soapdog** [comment]
bump! 

It is good that we can run the x86-32 version of VSCode on the Surface Pro X but if MS is serious about Windows on ARM then it needs to step up its support for development. One of the main reasons there are so few native arm64 apps for Windows is that there is no support for developing on arm64 machines. Everyone that spent ~2k on Surface Pro X are running win32 32bits emulated software or remoting into WSL. I've posted [about this](http://andregarzia.com/2020/01/windows-on-arm-needs-more-support-from-developers.html) in the past.

**Comment by ryan-allen** [comment]
@soapdog i've jumped onto the arm bandwagon with the galaxy book s, the battery life is awesome, and the LTE well and truly gives proper portability, it's a trade off but it's worth it if you can manage it.

i'm using x86 emulation with vscode, and using remote ssh to work with asp.net core on x64. it works well enough for day to day, and edge/firefox arm builds work really really well.

i was able to run an unofficial code oss build and it's lighting quick, but the extensions don't work yet.

wsl works great too, but i'm opting for using a slightly beefier virtual server instead via ssh remote.

once there's a arm64 build with ssh remote the experience will be wonderful, hopefully it's not too far off :)

**Comment by bkniffler** [comment]
@ryan-allen After another month, how happy are you? Could you give a quick update about your current setup and satisfaction with the performance? I'd be very happy to hear back, currently looking into ARM Windows for nodeJS/vscode. Cheers mate!

**Comment by ryan-allen** [comment]
@bkniffler hey bloke! i spent quite a while trying out a number of approaches, and i have settled on the following set up: i'm using (https://github.com/cdr/code-server)[code server] running on a x64 ubuntu server, i do all development through code-oss running remotely via this project. i use microsoft edge (which is native arm) to run code-server 'as an app' so that it captures key commands that are otherwise captured by the browser. there are features missing (e.g. debugging) because these are only available via vs-code released my msft. i'm using windows terminal preview which also has an arm build, so my day to day feels native and i have 95% of what i need with this setup, though it does require me to have an internet connection in order to work. 

it feels native, i use ssh port forwarding to hit local development servers on the virtual server i'm working on day to day. i'm very happy!

vscode running on the galaxy book s seemed sluggish due to is being 32bit emulated (this did work though, it just didn't feel 'fast').

one day if we get vscode arm and remote ssh it's going to be a winner. the battery life is unbelievable and built in LTE is amazing :)

**Comment by bkniffler** [comment]
@ryan-allen Thanks for the extensive feedback lad, it really seems incredible what Microsoft accomplished with the latest edge browser. I'm currently using an iPad with raspberry pi4 and code server, so I'm kinda familiar with the setup, but iPad just doesn't (yet?) feel like a real computer. Just to be sure, your x64 ubuntu server is physically on a different device? If yes, was it due to setup difficulty or performance benefits of a real server?

**Comment by davide-rota** [comment]
@bkniffler Hi, there's also [Visual Studio Online](https://visualstudio.microsoft.com/services/visual-studio-online/) that offers an alternative to code-server (now with docker support it's not bad)!

**Comment by ryan-allen** [comment]
@bkniffler i'm using a classic vps type thing to run code-server, and to develop on (vultr, ssd, cheaper and faster than aws/etc), it feels pretty much native as they have sydney edge locations. 

i have been working via remote-ssh for a couple of years as it allows me to work from different computers and have the same environment set up, i very much love remote-ssh!

i'm hoping visual studio online will allow remote-ssh to custom servers, @davide-rota does the docker support allow this yet? :)

**Comment by davide-rota** [comment]
@bkniffler you can have remote ssh and remote docker support to your own linux server right now with vscode remote extensions ( https://code.visualstudio.com/docs/remote/remote-overview ).
It becomes a bit more confusing when you want to combine remote docker+ssh as you could have remote docker over ssh **or** remote ssh and then use the docker extension work remotely.
Another complex scenario that works is to have docker-compose build remotely multiple containers and then attach vscode to the containers to develop and debug.
VS Online adds a managed/hosted environment and a vscode that can be used in the browser.

**Comment by soapdog** [comment]
@bkniffler I'm quite happy with the surface pro x but as [I blogged in the past](https://andregarzia.com/2020/01/windows-on-arm-needs-more-support-from-developers.html), I think Microsoft needs to setup it support for developers. There is basically no support for "dynamic languages" running native on Windows 10 on ARM, there is no released NodeJS aarch64 for Windows for example. There is an unofficial release but good luck using it and compiling something from source using `node-gyp`. Everything is almost in the right position for it to just work but it is not there yet. I'm using VSCode compiled for 32bits and I really wish I was able to run Electron apps as native aarch64 on this machine. Not because of the performance which is OK but because of battery life and making the Surface Pro X live to its full potential. I don't want to hijack or divert this issue, so focusing on VSCode, there are a gazilion developers out there that would jump on always-connected PCs if they'd run their dev setups, but the tools are not there yet.
